### PR TITLE
drivers: i2c: nrf5: Make the driver thread-safe

### PR DESCRIPTION
**Pull request by tkln**
Locking is added to externally callable API functions.

Signed-off-by: Aapo Vienamo <aapo.vienamo@iki.fi>

### CODE REVIEW

**Comment by tkln** [comment]
Could you please review this ￼@carlescufi @anangl @mnkp

File: drivers/i2c/i2c_nrf5.c
Comment (tbursztyka) [inline drivers/i2c/i2c_nrf5.c]: you could use K_MUTEX_INITIALIZER instead.

**Review by tbursztyka — CHANGES_REQUESTED** [review]
Why not using just a semaphore with 1 as limit? It's lighter.

**Review by carlescufi — APPROVED** [review]
Looks good

**Comment by carlescufi** [comment]
@tbursztyka can you re-review this?

File: drivers/i2c/i2c_nrf5.c
Comment (tbursztyka) [inline drivers/i2c/i2c_nrf5.c]: use K_SEM_DEFINE() in line 354 (for instance):

static struct i2c_nrf5_data i2c_nrf5_data_0 = {
K_SEM_DEFINE(lock, 1, 1),
};

I think it should work. 

Do the same for port 1

**Review by tbursztyka — CHANGES_REQUESTED** [review]

File: drivers/i2c/i2c_nrf5.c
Comment (carlescufi) [reply drivers/i2c/i2c_nrf5.c]: What's wrong with k_sem_init() I do the same as @tkln in other drivers

Comment (tbursztyka) [reply drivers/i2c/i2c_nrf5.c]: It's a tiny optimization to use K_SEM_DEFINE. I can dismiss my review, it's not a blocker after all.

**Review by tbursztyka — APPROVED** [review]

File: drivers/i2c/i2c_nrf5.c
Comment (tkln) [reply drivers/i2c/i2c_nrf5.c]: It's not wrong in the sense of being incorrect, but the semaphore does get initialized twice, once statically to zero and then properly by k_ksem_init(). Initializing it correctly statically has no runtime overhead and is probably preferrable.

Comment (carlescufi) [reply drivers/i2c/i2c_nrf5.c]: @tbursztyka understood. @tkln OK, feel free to update the PR before we merge it then

Comment (tkln) [reply drivers/i2c/i2c_nrf5.c]: @tbursztyka @carlescufi  Now that I tried to implement it, it seems that K_SEM_DEFINE() cannot be used with struct members. It looks like K_SEM_INITIALIZER() would do what we need, but it has apparently been deprecated.

Comment (tbursztyka) [reply drivers/i2c/i2c_nrf5.c]: You can use _K_SEM_INITIALIZER(), see commit 65a9d2a94a

Idea is to restrict usage of these macros to kernel space, and you driver is in kernel space so it's fine.

**Pull request merged**
